File: spec/components/alchemy/admin/page_select_spec.rb
require "rails_helper"

RSpec.describe Alchemy::Admin::PageSelect, type: :component do
  before do
    render
  end

  context "without parameters" do
    subject(:render) do
      render_inline(described_class.new) { "Page Select Content" }
    end

    it "should render the component and render given block content" do
      expect(page).to have_selector("alchemy-page-select")
      expect(page).to have_text("Page Select Content")
    end

    it "should not allow clearing" do
      expect(page).not_to have_selector("alchemy-page-select[allow-clear]")
    end

    it "should have the default placeholder" do
      expect(page).to have_selector("alchemy-page-select[placeholder='Search page']")
    end

    it "should have the default page api - url" do
      expect(page).to have_selector("alchemy-page-select[url='/api/pages']")
    end

    it "should not have a selection" do
      expect(page).to_not have_selector("alchemy-page-select[selection]")
    end
  end

  context "with page" do
    let(:alchemy_page) { create(:alchemy_page, id: 123, name: "Test Page") }
    subject(:render) do
      render_inline(described_class.new(alchemy_page))
    end

    it "should have a serialized page information" do
      expect(page).to have_selector('alchemy-page-select[selection="{\"id\":123,\"name\":\"Test Page\",\"url_path\":\"/test-page\"}"]')
    end
  end

  context "with url" do
    subject(:render) do
      render_inline(described_class.new(nil, url: "/foo-bar"))
    end

    it "should have an url parameter" do
      expect(page).to have_selector('alchemy-page-select[url="/foo-bar"]')
    end
  end

  context "with allow clear" do
    subject(:render) do
      render_inline(described_class.new(nil, allow_clear: true))
    end

    it "should not have a allow_clear attribute" do
      expect(page).to have_selector("alchemy-page-select[allow-clear]")
    end
  end

  context "with custom placeholder" do
    subject(:render) do
      render_inline(described_class.new(nil, placeholder: "Custom Placeholder"))
    end

    it "should have a custom placeholder" do
      expect(page).to have_selector("alchemy-page-select[placeholder='Custom Placeholder']")
    end
  end

  context "with query parameter" do
    subject(:render) do
      render_inline(described_class.new(nil, query_params: {foo: :bar}))
    end

    it "should have serialized custom parameter" do
      expect(page).to have_selector('alchemy-page-select[query-params="{\"foo\":\"bar\"}"]')
    end
  end
end

File: spec/javascript/alchemy_admin/components/page_select.spec.js
import { renderComponent } from "./component.helper"
import "alchemy_admin/components/page_select"

const PAYLOAD = "<img src=x onerror=alert(1)>"

describe("alchemy-page-select", () => {
  /**
   *
   * @type {HTMLElement | undefined}
   */
  let component = undefined

  describe("without configuration", () => {
    beforeEach(() => {
      const html = `
        <alchemy-page-select>
          <input type="text">
        </alchemy-page-select>
      `
      component = renderComponent("alchemy-page-select", html)
    })

    it("should render the input field", () => {
      expect(component.getElementsByTagName("input")[0]).toBeInstanceOf(
        HTMLElement
      )
    })

    it("should initialize Select2", () => {
      expect(
        component.getElementsByClassName("select2-container").length
      ).toEqual(1)
    })

    it("should not show a remove 'button'", () => {
      expect(
        document.querySelector(".select2-container.select2-allowclear")
      ).toBeNull()
    })
  })

  describe("allow clear", () => {
    beforeEach(() => {
      const html = `
        <alchemy-page-select allow-clear>
          <input type="text">
        </alchemy-page-select>
      `
      component = renderComponent("alchemy-page-select", html)
    })

    it("should show a remove 'button'", () => {
      expect(component.allowClear).toBeTruthy()
    })
  })

  describe("query params", () => {
    beforeEach(() => {
      const html = `
        <alchemy-page-select query-params="{&quot;foo&quot;:&quot;bar&quot;}">
          <input type="text">
        </alchemy-page-select>
      `
      component = renderComponent("alchemy-page-select", html)
    })

    it("should receive query parameter", () => {
      expect(JSON.parse(component.queryParams)).toEqual({ foo: "bar" })
    })

    it("should add the query parameter to the API call", () => {
      expect(component.ajaxConfig.data("test").q.foo).toEqual("bar")
    })
  })

  describe("rendering", () => {
    const page = (attrs) => ({
      name: "Home",
      site: { name: "Default site" },
      url_path: "/home",
      language_code: "en",
      ...attrs
    })

    const render = (attrs, term = "") => {
      const el = document.createElement("div")
      el.innerHTML = component._renderListEntry(page(attrs), term)
      return el
    }

    beforeEach(() => {
      const html = `
        <alchemy-page-select>
          <input type="text">
        </alchemy-page-select>
      `
      component = renderComponent("alchemy-page-select", html)
    })

    it("escapes the page name", () => {
      const name = `Home ${PAYLOAD}`
      const el = render({ name }, "Home")
      expect(el.querySelector("img")).toBeNull()
      expect(
        el.querySelector(".page-select--page-name").textContent.trim()
      ).toEqual(name)
    })

    it("escapes the site name", () => {
      const el = render({ site: { name: PAYLOAD } })
      expect(el.querySelector("img")).toBeNull()
      expect(
        el.querySelector(".page-select--site-name").textContent.trim()
      ).toEqual(PAYLOAD)
    })

    it("escapes the url path", () => {
      const el = render({ url_path: PAYLOAD })
      expect(el.querySelector("img")).toBeNull()
      expect(
        el.querySelector(".page-select--page-urlname").textContent.trim()
      ).toEqual(PAYLOAD)
    })

    it("escapes the language code", () => {
      const el = render({ language_code: PAYLOAD })
      expect(el.querySelector("img")).toBeNull()
      expect(
        el.querySelector(".page-select--language-code").textContent.trim()
      ).toEqual(PAYLOAD)
    })

    it("escapes the name of the selected page", () => {
      const el = document.createElement("div")
      el.innerHTML = component._renderResult(page({ name: PAYLOAD }))
      expect(el.querySelector("img")).toBeNull()
      expect(el.textContent).toEqual(PAYLOAD)
    })
  })
})
